avfilter: Added text2bitmap subtitle filter

tcoza opened the pull request:
Added a text2bitmap subtitle filter which converts text-based subtitle tracks
to bitmap-based subtitle tracks. The filter uses libass to render the subtitles.
It takes as parameters an output height and width, as well as a number of colors
in the output palette as well as sources of fonts. All its arguments are optional.

softworkz committed:
avcodec/dvbsubdec: Fix conditions for fallback to default resolution

The previous code expected a segment of type CLUT definition to exist
in order to accept a set of segments to be complete.
This was an incorrect assumption as the presence of a CLUT segment
is not mandatory.
(version 1.6.1 of the spec is probably a bit more clear about this
than earlier versions: https://www.etsi.org/deliver/etsi_en/
300700_300799/300743/01.06.01_20/en_300743v010601a.pdf)

The flawed condition prevented proper fallback to using the default
resolution for the decoding context.

Signed-off-by: softworkz <softworkz@hotmail.com>

softworkz committed:
avcodec/subtitles: Migrate subtitle encoders to frame-based API

and provide a compatibility shim for the legacy api

Signed-off-by: softworkz <softworkz@hotmail.com>

tcoza committed:
avfilter/text2graphicsub: Added text2graphicsub subtitle filter

Added a text2graphicsub subtitle filter which converts text-based subtitle tracks
to bitmap-based subtitle tracks. The filter uses libass to render the subtitles.
It takes as parameters an output height and width, as well as a number of colors
in the output palette as well as sources of fonts. All its arguments are optional.

Signed-off-by: tcoza <traian.coza@gmail.com>

softworkz commented on configure:
```diff
@@ -3730,6 +3730,7 @@ sr_filter_select="dnn"
 stereo3d_filter_deps="gpl"
 splitcc_filter_deps="avcodec"
 subscale_filter_deps="swscale avcodec"
+text2bitmap_filter_deps="libass"
```
Needs to be added in alphabetical order.
But it seems I'll need to fix my own entries as well. The frequent rebasing has mixed up the order.

softworkz replied:
You are using libavformat and libavcodec, so you'll need to add them just like `subtitles_filter_deps`

softworkz commented on configure:
```diff
@@ -7866,7 +7867,8 @@ print_enabled_components(){
-    echo "static const $struct_name * const $name[] = {" > $TMPH
+    echo "/* Automatically generated by configure - do not modify! */" > $TMPH
+    echo "static const $struct_name * const $name[] = {" >> $TMPH
     for c in $*; do
```
Why?
This is none of our business :-)

tcoza replied:
It just bothered me, I had just modified an auto-generated file without knowing it was auto-generated, and I wondered why it didn't have that usual disclaimer, so I decided to fix it. But I just remembered FFmpeg standards would have this be a separate commit, so I'll remove it 👍

softworkz replied:
I just made a test to understand which files this is actually affecting and I agree that this change makes a lot of sense and improves consistency among all generated files. It's a good idea to submit it in a separate patch.

softworkz commented on libavfilter/sf_text2bitmap.c:
```diff
@@ -0,0 +1,539 @@
+        { "height",     "output height",           OFFSET(height),     AV_OPT_TYPE_INT,    .min = 0, .max = 10000, .default_val.i64 = 0,    .flags = FLAGS },
+        { "n",          "number of output colors", OFFSET(num_colors), AV_OPT_TYPE_INT,    .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
+        { "num_colors", "number of output colors", OFFSET(num_colors), AV_OPT_TYPE_INT,    .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
+        { "f",          "fonts container",         OFFSET(fonts),      AV_OPT_TYPE_STRING,                         .default_val.str = NULL, .flags = FLAGS },
+        { "fonts",      "fonts container",         OFFSET(fonts),      AV_OPT_TYPE_STRING,                         .default_val.str = NULL, .flags = FLAGS },
```
The naming is a bit weird. Nobody knows what a "fonts container" is. Also 'fonts' doesn't sound like you would need to specify a filename to media file containing font attachments.
I would just just copy the options from vf_subtitles:

https://github.com/softworkz/FFmpeg/blob/34aee50a7f5a4b36f9f7daf827422d345c7868ed/libavfilter/vf_subtitles.c#L69-L72

and change the description `set the filename of file to read` to something like `media file to load fonts from`

softworkz commented on libavfilter/sf_text2bitmap.c:
```diff
@@ -0,0 +1,539 @@
+#define FLAGS (AV_OPT_FLAG_SUBTITLE_PARAM|AV_OPT_FLAG_FILTERING_PARAM)
+
+static const AVOption text2bitmap_options[] = {
+        { "w",          "output width",            OFFSET(width),      AV_OPT_TYPE_INT,    .min = 0, .max = 10000, .default_val.i64 = 0,    .flags = FLAGS },
+        { "width",      "output width",            OFFSET(width),      AV_OPT_TYPE_INT,    .min = 0, .max = 10000, .default_val.i64 = 0,    .flags = FLAGS },
+        { "h",          "output height",           OFFSET(height),     AV_OPT_TYPE_INT,    .min = 0, .max = 10000, .default_val.i64 = 0,    .flags = FLAGS },
+        { "height",     "output height",           OFFSET(height),     AV_OPT_TYPE_INT,    .min = 0, .max = 10000, .default_val.i64 = 0,    .flags = FLAGS },
```
A parameter of type AV_OPT_TYPE_IMAGE_SIZE would be better.

softworkz commented on libavfilter/sf_text2bitmap.c:
```diff
@@ -0,0 +1,539 @@
+        .priv_class      = &text2bitmap_class,
+        FILTER_INPUTS(inputs),
+        FILTER_OUTPUTS(outputs),
+        FILTER_SUBFMTS(AV_SUBTITLE_FMT_TEXT, AV_SUBTITLE_FMT_ASS)
```
Drop AV_SUBTITLE_FMT_TEXT

softworkz reviewed (changes requested):
This was from a quick look, I didn't look at the code closely yet.
Thanks

softworkz commented on libavfilter/sf_text2bitmap.c:
```diff
@@ -0,0 +1,539 @@
+        .priv_class      = &text2bitmap_class,
+        FILTER_INPUTS(inputs),
+        FILTER_OUTPUTS(outputs),
+        FILTER_SUBFMTS(AV_SUBTITLE_FMT_TEXT, AV_SUBTITLE_FMT_ASS)
```
You cannot use FILTER_SUBFMTS(). This is only for filters where input and output are of the same format.
You need to use the query_formats method instead.

tcoza replied:
Can you explain what that function is supposed to do? I looked as some examples but I don't understand.

softworkz replied:
This is about "format negotiation". A filter can support multiple combinations of input and output formats (not media types). The filtering implementation uses this to figure out a viable combination over a longer chain of filters.

In your case, you support only  AV_SUBTITLE_FMT_ASS and input and  AV_SUBTITLE_FMT_BITMAP as output.

You can look at the OCR filter, which is very similar just with the formats reversed.

At the bottom you need this:

https://github.com/softworkz/FFmpeg/blob/a66debd96ef9593a628ac7830819f345a98bdb02/libavfilter/sf_graphicsub2text.c#L1131

and then the same query_formats() function just with the formats reversed.

softworkz commented on libavfilter/sf_text2bitmap.c:
```diff
@@ -0,0 +1,539 @@
+    image_rgba_size = nb_pixels * 4 * sizeof(uint8_t);
+    image_rgba = (uint8_t *)malloc(image_rgba_size);
+    memset(image_rgba, 0, image_rgba_size);        // Set all to transparent
+    for (; image != NULL; image = image->next)
+    {
+        // Set pixels
+        for (int y = 0; y < image->h; y++)
+            for (int x = 0; x < image->w; x++) {
+                int x_area = image->dst_x + x - area->x;
+                int y_area = image->dst_y + y - area->y;
+                uint8_t alpha_b = image_rgba[4 * (y_area * area->w + x_area) + 3];
+                uint8_t alpha_a = AA(image->color) * image->bitmap[y * image->stride + x] >> 8;
+                uint8_t alpha = (ALPHA_MAX * alpha_a + alpha_b * (ALPHA_MAX - alpha_a)) / ALPHA_MAX;
+                image_rgba[4 * (y_area * area->w + x_area) + 3] = alpha;
+                if (alpha == 0) continue;       // The components don't matter (plus we get FPE otherwise)
+                for (int c = 0; c < 3; c++) {
+                    uint8_t comp_b = image_rgba[4 * (y_area * area->w + x_area) + c];
+                    uint8_t comp_a = (image->color >> ((c + 1) * 8)) & 0xFF;
+                    uint8_t comp = (comp_a * alpha_a * ALPHA_MAX + comp_b * alpha_b * (ALPHA_MAX - alpha_a)) / (alpha * ALPHA_MAX);
+                    image_rgba[4 * (y_area * area->w + x_area) + c] = comp;
+                }
+            }
+    }
```
Why don't you use ff_draw_ like the other implementations?

tcoza replied:
First time I hear of it, thanks!

softworkz replied:
> First time I hear of it, thanks!

First time?? - What I mean is how vf_subtitles and vf_overlaytextsubs are doing it..

Don't forget the ff_draw_init() call for transparency handling (like textsub2video does)

softworkz commented on libavfilter/sf_text2bitmap.c:
```diff
@@ -0,0 +1,539 @@
+    // Create rgba image
+    nb_pixels = area->w * area->h;
+    image_rgba_size = nb_pixels * 4 * sizeof(uint8_t);
+    image_rgba = (uint8_t *)malloc(image_rgba_size);
```
malloc is not allowed. Use one of the av_ functions for allocating and freeing
Edit: precisely av_malloc() or  av_mallocz() which also zeroes memory.

Before very recently I couldn't have told you a good reason, but one (probably there are others) is that it is possible that each avlib is linked to a separate C runtime, which means that you would get a crash when freeing something in one lib which was alloced in another lib,

softworkz reviewed (changes requested):
Some more comments

softworkz commented on libavfilter/sf_text2bitmap.c:
```diff
@@ -0,0 +1,539 @@
+                    &area->pal[0], context->num_colors);
+
+    // Clean up
+    free(image_rgba);
```
av_free()

softworkz commented on libavfilter/sf_text2bitmap.c:
```diff
@@ -0,0 +1,539 @@
+    av_freep(&context->codeword);
+    av_freep(&context->codeword_closest_codebook_idxs);
+    avpriv_elbg_free(&context->elbg);
+    free(context);
```
av_free()

softworkz commented on libavfilter/sf_text2bitmap.c:
```diff
@@ -0,0 +1,539 @@
+        track->n_events = track->max_events = 1;
+        track->events[0].Start = frame->subtitle_timing.start_pts / (AV_TIME_BASE / 1000);
+        track->events[0].Duration = frame->subtitle_timing.duration / (AV_TIME_BASE / 1000);
+        track->events[0].Effect = strdup(dialog->effect);
```
av_strdup()

tcoza commented:
Ok, I made the modifications suggested. I forced pushed to the same commit, I'm not sure this is how it works, I just am used to adding to the same commit because that's how ffmpeg wants it.

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,578 @@
+         * for most frames but noticeably dissimilar for frames with
+         * a lot of semi-transparent parts. I am not sure which
+         * of the results is the correct one. */
+        // Set pixels
```
The ffmpeg devs don't accept submissions with commented-out code.
I do this a lot in my own projects, especially when I think there might be some value in it that might get lost, but well - in a huge project like ffmpeg it would probably cause a lot of bloat.

When you say there's a difference in alpha processing when you compare both implementation, you could try to init ff_draw without the FF_DRAW_PROCESS_ALPHA flag.

IIRC the purpose of this flag is about how the alpha component of the draw-colors should affect the alpha of the target bitmap. The way I've used it so far is:

When drawing onto a video frame with content, I don't use FF_DRAW_PROCESS_ALPHA, to keep the target frame fully opaque. If drawing onto a transparent image, I use FF_DRAW_PROCESS_ALPHA, because otherwise the RGB values would change but all pixels would remain fully transparent.
(IIRC - it's been a while..)

tcoza replied:
Yeah, I tried not setting that flag, it gives the same result. Too bad.

softworkz replied:
This would normally mean that none of your palette colors has an intermediate (somewhere in the middle between 0 and 255) alpha value?

tcoza replied:
Hmm no but it does. As I said, the parts that differ are semi-transparent. The semi=transparent parts are just different. I can show you examples if you want.

softworkz replied:
Let's look at an example. Assume there's a palette color RGBA 100, 100, 100, 100.
What does the resulting (non-paletted) bitmap have as color at these pixels (in either case)?

tcoza replied:
Not sure how I should do this. I was thinking more of sending you screenshots of a subtitle that renders differently in both cases. I can do that a little bit later tho.

softworkz replied:
Another way to test this is to compare the following:

- subtitles burnt in via overlaytextsubs
- subtitles converted with your text2graphicsub and then burnt in via overlaygraphicsubs

softworkz commented on doc/filters.texi:
```diff
@@ -27633,6 +27633,42 @@ ffmpeg -y -y -loglevel verbose -i "https://streams.videolan.org/samples/sub/larg
+rendering the subtitles. Can be used with 'fonts'. Can be omitted.
+
+@end table
+
```
A working example command line would be great

softworkz replied:
Thanks for adding

softworkz committed:
avfilter/snull,strim: Add snull and strim filters

Signed-off-by: softworkz <softworkz@hotmail.com>

softworkz reviewed (changes requested):
Thanks for the updates. Getting better and better :-)

One missing bit is a signed-off line at the bottom of the commit message.
You don't need to really sign the commits, just add this line,

softworkz commented:
I have pushed a new version which:

- has all entries in allfilters and makefile in correct alphabetical order
- adds snull and strim filters
- supports simple filter strings for subtitles from the command line

Once you have rebased and updated, I will apply your code locally and do some testing.
Then we can submit a new patchset version with your filter included.

tcoza commented:
This rebase is driving me insane. There are about 100 merge conflicts and I have no idea how to solve them. I am just solving at random.

softworkz commented:
I know what you mean, I've been there.

Shall I do it for you?

softworkz commented:
But you have a single commit only - that's not even close to the core of rebasing hell... :-)

tcoza commented:
Sure, if you can. It seems like I should keep the commits from your update, but that seems to be deleting a lot of documentation. I don't understand it.

softworkz commented:
Just 

- duplicate your branch to a backup branch
- then reset your working branch to latest upstream (my branch) state
- then cherry-pick your commit from the backup branch

You'll get a simple conflict resolution scenario then

tcoza commented:
Cherry-pick? Is that a technical term? What does it mean?

softworkz commented:
> Sure, if you can. It seems like I should keep the commits from your update, but that seems to be deleting a lot of documentation. I don't understand it.

Ah, you merged:

![image](https://user-images.githubusercontent.com/4985349/173491068-ccaf2de0-f4e5-4f02-a016-6dce11166875.png)

You must not merge, you need to rebase when pulling

tcoza commented:
I know, but github only does merge, so I merge, and then I just do the rebasing on my computer. Unless there's a way to just do rebase fom Github, but I'm unaware of it.

softworkz commented:
> Cherry-pick? Is that a technical term? What does it mean?

Yes, it's a GIT term, but in fact you just need to rebase instead of merge when pulling.

So, forget my previous instructions. Instead, reset you branch to your original commit.

Then pull with rebase. Which GIT client do you use?

tcoza commented:
I just use git for linux

softworkz commented:
I've sent you a rebased version of your path.

Now, you need to reset your working branch to my most recent commit: 251582a8525167c89d8b5f51f7c8d9bf959ae7a8

Then you can apply your patch that I just setn on top of that via

`git am <xx.patch>`

tcoza commented:
I did the cherry-pick thing, do you think that's ok? It was very easy..

softworkz commented:
> I did the cherry-pick thing, do you think that's ok? It was very easy..

When you had exactly two conflicts (in allfilters and makefile) to solve, then you did it right and you can now force-push your branch from local to your GitHub repo

tcoza commented:
Okay, yes I just had 2 conflicts. Thanks anyway. How come rebase is so complicated compared to cherry-pick? Isn't it basically the same thing? At least in my case.

softworkz commented:
For testing, you can still try to apply the patch I sent you with `git am`. When it says "nothing to commit", it's all ok.

tcoza commented:
I'll try that.

softworkz commented:
> Okay, yes I just had 2 conflicts. Thanks anyway. How come rebase is so complicated compared to cherry-pick? Isn't it basically the same thing? At least in my case.

rebase is not complicated, it's just `git rebase ...`.

The difference is that with git rebase, it tries to apply your commits on top of the updated base branch, while with merge, it tries to integrate the updated branch with commits that come AFTER your changes. 
But you when you do so in this case, you already have my original commits, and now you start applying all my commits again. 
That means that the first commit already creates chaos because it's being applied on the end result of my previous completion state.

tcoza commented:
All right, here we go, everything should be good.

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,492 @@
+        [5] = AV_LOG_INFO,      /* <undefined> */
+        [6] = AV_LOG_VERBOSE,   /* MSGL_V */
+        [7] = AV_LOG_DEBUG,     /* MSGL_DBG2 */
+};
```
Indentation invalid (wrt ffmpeg code style)

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,492 @@
+        "application/vnd.ms-opentype",
+        "application/x-font-ttf",
+        NULL
+};
```
Indentation invalid (wrt ffmpeg code style)

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,492 @@
+    avpriv_ass_split_free(context->ass_split_context);
+    ass_free_track(context->track);
+    ass_renderer_done(context->renderer);
+    ass_library_done(context->library);
```
Causes segfaults when called with null pointers. Why didn't you just copy over the code:

```
    if (context->track)
        ass_free_track(context->track);
    if (context->renderer)
        ass_renderer_done(context->renderer);
    if (context->library)
        ass_library_done(context->library);

    context->track = NULL;
    context->renderer = NULL;
    context->library = NULL;
```

If you want to reproduce, just try an invalid command line, like:

`ffmpegd.exe -y -i  INPUT -map 0:v -map 0:999 -filter_complex "[0:s]text2graphicsub[dvd_sub]" -map [dvd_sub] -c copy -c:s dvd_subtitle output.mkv`

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,492 @@
+    char *filename, *fontsdir;
+} ASS_Context;
+
+static void ass_log(int ass_level, const char *fmt, va_list args, void *ctx);
+static void ass_image_to_area_palletization(ASS_Context *context, ASS_Image *image, AVSubtitleArea *area);
+static void init_palettizecontext(PalettizeContext **palettize_context);
+static void free_palettizecontext(PalettizeContext **palettize_context);
+static int stream_is_font(AVStream * st);
```
Forward declarations without need are disliked. Please just order the functions accordingly.
(I don't object this personally)

tcoza replied:
What do you mean without need? They are all needed there.

softworkz replied:
> Please just order the functions accordingly.

They are not needed. This is your filter code without forward declarations:

```
/*
 * Copyright (c) 2021 tcoza
 *
 * This file is part of FFmpeg.
 *
 * FFmpeg is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * FFmpeg is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with FFmpeg; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

/**
 * @file
 * convert text subtitles to bitmap subtitles filter
 */

#include <ass/ass.h>
#include "libavutil/avstring.h"
#include "libavutil/ass_split_internal.h"

#include "internal.h"
#include "avfilter.h"
#include "drawutils.h"
#include "libavutil/opt.h"
#include "libavutil/buffer.h"
#include "libavutil/internal.h"
#include "libavformat/avformat.h"
#include "libavcodec/elbg.h"
#include "libavutil/lfg.h"

typedef struct PalettizeContext {
    int *codeword, *codebook;
    int *codeword_closest_codebook_idxs;
    int r_idx, g_idx, b_idx, a_idx;
    struct ELBGContext *elbg;
    AVLFG lfg;
} PalettizeContext;

typedef struct ASS_Context {
    ASS_Library *library;
    ASS_Renderer *renderer;
    ASS_Track *track;
    ASSSplitContext *ass_split_context;
    PalettizeContext *palettize_context;
    FFDrawContext draw_context;
    struct { int width; int height; } size;
    int num_colors;
    char *filename, *fontsdir;
} ASS_Context;

/* libass supports a log level ranging from 0 to 7 */
static const int ass_libavfilter_log_level_map[] = {
    [0] = AV_LOG_FATAL,     /* MSGL_FATAL */
    [1] = AV_LOG_ERROR,     /* MSGL_ERR */
    [2] = AV_LOG_WARNING,   /* MSGL_WARN */
    [3] = AV_LOG_WARNING,   /* <undefined> */
    [4] = AV_LOG_INFO,      /* MSGL_INFO */
    [5] = AV_LOG_INFO,      /* <undefined> */
    [6] = AV_LOG_VERBOSE,   /* MSGL_V */
    [7] = AV_LOG_DEBUG,     /* MSGL_DBG2 */
};

static void ass_log(int ass_level, const char *fmt, va_list args, void *ctx)
{
    const int ass_level_clip = av_clip(ass_level, 0,
        FF_ARRAY_ELEMS(ass_libavfilter_log_level_map) - 1);
    const int level = ass_libavfilter_log_level_map[ass_level_clip];
    av_vlog(NULL, level, fmt, args);
    av_log(NULL, level, "\n");
}

static void set_area_bounds(ASS_Image *image, AVSubtitleArea *area)
{
    ASS_Image *image_o = image;
    area->x = image->dst_x + 1; area->w = -1;
    area->y = image->dst_y + 1; area->h = -1;
    for (image = image_o; image != NULL; image = image->next)
    {
        // Set image position to encompass all images
        if (image->dst_x < area->x) area->x = image->dst_x;
        if (image->dst_y < area->y) area->y = image->dst_y;
    }
    for (image = image_o; image != NULL; image = image->next)
    {
        // Set image size to encompass all images
        if (image->dst_x + image->w > area->x + area->w)
            area->w = image->dst_x + image->w - area->x;
        if (image->dst_y + image->h > area->y + area->h)
            area->h = image->dst_y + image->h - area->y;
    }
    area->linesize[0] = area->w;            // How it is
}

#define ALPHA_THRESHOLD 0b10000000
#define ALPHA_MAX ((1 << (sizeof(uint8_t) << 3)) - 1)
static void palettize_image(PalettizeContext *const s,
                            const int w, const int h,
                            uint8_t *src_data, const int src_linesize,
                            uint8_t *dst_data, const int dst_linesize,
                            uint32_t *dst_pal, const int num_output_colors);

/* libass stores an RGBA color in the format RRGGBBTT, where TT is the transparency level */
#define AR(c)  (((c)>>24)&0xFF)
#define AG(c)  (((c)>>16)&0xFF)
#define AB(c)  (((c)>> 8)&0xFF)
#define AA(c)  ((0xFF-(c))&0xFF)
#define TO_AREA_COLOR(c) (AB(c)|(AG(c)<<8)|(AR(c)<<16)|(AA(c)<<24))

static void ass_image_to_area_palletization(ASS_Context *context, ASS_Image *image, AVSubtitleArea *area)
{
    unsigned nb_pixels;
    size_t image_rgba_size;
    uint8_t *image_rgba;

    set_area_bounds(image, area);

    // Create rgba image
    nb_pixels = area->w * area->h;
    image_rgba_size = nb_pixels * 4 * sizeof(uint8_t);
    image_rgba = (uint8_t *)av_mallocz(image_rgba_size);
    for (; image != NULL; image = image->next)
    {
        uint8_t rgba_color[] = {AR(image->color), AG(image->color), AB(image->color), AA(image->color)};
        int linesize = area->w * 4;
        FFDrawColor color;
        ff_draw_color(&context->draw_context, &color, rgba_color);
        ff_blend_mask(&context->draw_context, &color,
                      &image_rgba, &linesize, area->w, area->h,
                      image->bitmap, image->stride, image->w, image->h,
                      3, 0, image->dst_x - area->x, image->dst_y - area->y);
    }

    area->nb_colors = context->num_colors;
    area->buf[0] = av_buffer_alloc(nb_pixels * sizeof(uint8_t));

    palettize_image(context->palettize_context,
                    area->w, area->h,
                    image_rgba, area->w * 4,
                    area->buf[0]->data, area->w,
                    &area->pal[0], context->num_colors);

    // Clean up
    av_free(image_rgba);
}

static void init_palettizecontext(PalettizeContext **palettizecontext)
{
    uint8_t rgba_map[4];
    PalettizeContext *context = (PalettizeContext *)av_malloc(sizeof(PalettizeContext));
    context->codebook = NULL;
    context->codeword = NULL;
    context->codeword_closest_codebook_idxs = NULL;
    context->elbg = NULL;
    av_lfg_init(&context->lfg, 0xACBADF);
    ff_fill_rgba_map(&rgba_map[0], AV_PIX_FMT_RGB32);
    context->r_idx = rgba_map[0]; // R
    context->g_idx = rgba_map[1]; // G
    context->b_idx = rgba_map[2]; // B
    context->a_idx = rgba_map[3]; // A
    *palettizecontext = context;
}

static void free_palettizecontext(PalettizeContext **palettizecontext)
{
    PalettizeContext *context = *palettizecontext;
    av_freep(&context->codebook);
    av_freep(&context->codeword);
    av_freep(&context->codeword_closest_codebook_idxs);
    avpriv_elbg_free(&context->elbg);
    av_free(context);
    *palettizecontext = NULL;
}

static const char * const font_mimetypes[] = {
        "font/ttf",
        "font/otf",
        "font/sfnt",
        "font/woff",
        "font/woff2",
        "application/font-sfnt",
        "application/font-woff",
        "application/x-truetype-font",
        "application/vnd.ms-opentype",
        "application/x-font-ttf",
        NULL
};

static int stream_is_font(AVStream* st)
{
    const AVDictionaryEntry *tag = NULL;
    int n;

    if (st->codecpar->codec_type != AVMEDIA_TYPE_ATTACHMENT)
        return 0;

    tag = av_dict_get(st->metadata, "mimetype", NULL, AV_DICT_MATCH_CASE);

    if (tag) {
        for (n = 0; font_mimetypes[n]; n++) {
            if (av_strcasecmp(font_mimetypes[n], tag->value) == 0)
                return 1;
        }
    }
    return 0;
}

static av_cold int init(AVFilterContext *ctx)
{
    ASS_Context *context = ctx->priv;

    context->library = ass_library_init();
    ass_set_message_cb(context->library, ass_log, NULL);
    ass_set_fonts_dir(context->library, context->fontsdir);
    ass_set_extract_fonts(context->library, 1);

    if (context->filename) {
        AVFormatContext *video = NULL;
        int ret = avformat_open_input(&video, context->filename, NULL, NULL);
        if (ret < 0) {
            av_log(ctx, AV_LOG_ERROR, "Unable to open %s\n", context->filename);
            return ret;
        }
        for (int i = 0; i < video->nb_streams; i++) {
            const AVDictionaryEntry *tag;
            AVStream *st = video->streams[i];
            if (!stream_is_font(st)) continue;
            tag = av_dict_get(st->metadata, "filename", NULL, AV_DICT_MATCH_CASE);
            if (!tag) continue;
            av_log(NULL, AV_LOG_DEBUG, "Loading attached font: %s\n", tag->value);
            ass_add_font(context->library, tag->value,
                 (char *)st->codecpar->extradata,
                 st->codecpar->extradata_size);
        }
        avformat_close_input(&video);
    }

    context->renderer = ass_renderer_init(context->library);
    ass_set_pixel_aspect(context->renderer, 1);
    ass_set_shaper(context->renderer, 0);
    ass_set_fonts(context->renderer, NULL, NULL, 1, NULL, 1);

    context->track = NULL;              // Will be init'd in filter_frame
    context->ass_split_context = NULL;  //

    init_palettizecontext(&context->palettize_context);

    ff_draw_init(&context->draw_context, AV_PIX_FMT_BGRA, FF_DRAW_PROCESS_ALPHA);

    return 0;
}

static int config_input(AVFilterLink *inlink)
{
    const AVFilterContext *ctx = inlink->dst;
    ASS_Context *context = ctx->priv;
    if (!context->size.width)  context->size.width  = inlink->w;
    if (!context->size.height) context->size.height = inlink->h;
    if (!context->size.height || !context->size.width) {
        av_log(NULL, AV_LOG_ERROR, "A positive height and width are required to render subtitles\n");
        return AVERROR_EXIT;
    }
    ass_set_frame_size(context->renderer, context->size.width, context->size.height);
    ass_set_storage_size(context->renderer, context->size.width, context->size.height);

    return 0;
}

static int config_output(AVFilterLink *outlink)
{
    AVFilterContext *ctx = outlink->src;
    ASS_Context *context = ctx->priv;

    outlink->w = context->size.width;
    outlink->h = context->size.height;

    return 0;
}

static int filter_frame(AVFilterLink *inlink, AVFrame *frame)
{
    const AVFilterContext *ctx = inlink->dst;
    ASS_Context *context = ctx->priv;
    AVFilterLink *outlink = ctx->outputs[0];
    if (!context->track) {      // Init track
        if (frame->subtitle_header)
            context->track = ass_read_memory(context->library,
                (char *) frame->subtitle_header->data,
                frame->subtitle_header->size, NULL);
        else context->track = ass_new_track(context->library);
        if (!frame->subtitle_header) context->ass_split_context = avpriv_ass_split("");
        else context->ass_split_context = avpriv_ass_split((char *) frame->subtitle_header->data);
    }

    for (int r = 0; r < frame->num_subtitle_areas; r++)
    {
        ASS_Image *image;
        ASSDialog *dialog;
        ASS_Track *track = context->track;
        AVSubtitleArea *area = frame->subtitle_areas[r];
        if (area->type == SUBTITLE_BITMAP) continue;

        dialog = avpriv_ass_split_dialog(context->ass_split_context, area->ass);
        if (track->n_events > 0)
            ass_free_event(track, 0);
        track->n_events = 0;
        ass_alloc_event(track);
        track->n_events = track->max_events = 1;
        track->events[0].Start = frame->subtitle_timing.start_pts / (AV_TIME_BASE / 1000);
        track->events[0].Duration = frame->subtitle_timing.duration / (AV_TIME_BASE / 1000);
        track->events[0].Effect = strdup(dialog->effect);
        track->events[0].Layer = dialog->layer;
        track->events[0].MarginL = dialog->margin_l;
        track->events[0].MarginR = dialog->margin_r;
        track->events[0].MarginV = dialog->margin_v;
        track->events[0].Name = strdup(dialog->name);
        track->events[0].Text = strdup(dialog->text);
        track->events[0].ReadOrder = dialog->readorder;
        track->events[0].Style = 0;
        for (int style = 0; style < track->n_styles; style++)
            if (!strcmp(track->styles[style].Name, dialog->style))
                track->events[0].Style = style;
        track->events[0].render_priv = NULL;
        avpriv_ass_free_dialog(&dialog);

        image = ass_render_frame(context->renderer, track,
            track->events[0].Start + track->events[0].Duration / 2, NULL);
        if (image == NULL) {
            av_log(NULL, AV_LOG_WARNING, "failed to render ass: %s\n", area->ass);
            return 0;
        }
        ass_image_to_area_palletization(context, image, area);
        area->type = SUBTITLE_BITMAP;

        av_log(NULL, AV_LOG_DEBUG, "successfully rendered ass: %s\n", area->ass);
    }

    return ff_filter_frame(outlink, frame);
}

static av_cold void uninit(AVFilterContext *ctx)
{
    ASS_Context *context = ctx->priv;
    free_palettizecontext(&context->palettize_context);
    avpriv_ass_split_free(context->ass_split_context);

    if (context->track)
        ass_free_track(context->track);
    if (context->renderer)
        ass_renderer_done(context->renderer);
    if (context->library)
        ass_library_done(context->library);

    context->track = NULL;
    context->renderer = NULL;
    context->library = NULL;
}

// Copied from sf_graphicsub2text, with formats swapped
static int query_formats(AVFilterContext *ctx)
{
    AVFilterFormats *formats, *formats2;
    AVFilterLink *inlink = ctx->inputs[0];
    AVFilterLink *outlink = ctx->outputs[0];
    static const enum AVSubtitleType in_fmts[] = { AV_SUBTITLE_FMT_ASS, AV_SUBTITLE_FMT_NONE };
    static const enum AVSubtitleType out_fmts[] = { AV_SUBTITLE_FMT_BITMAP, AV_SUBTITLE_FMT_NONE };
    int ret;

    /* set input format */
    formats = ff_make_format_list(in_fmts);
    if ((ret = ff_formats_ref(formats, &inlink->outcfg.formats)) < 0)
        return ret;

    /* set output format */
    formats2 = ff_make_format_list(out_fmts);
    if ((ret = ff_formats_ref(formats2, &outlink->incfg.formats)) < 0)
        return ret;

    return 0;
}

static void palettize_image(PalettizeContext *const s,
                           const int w, const int h,
                           uint8_t *src_data, const int src_linesize,
                           uint8_t *dst_data, const int dst_linesize,
                           uint32_t *dst_pal, const int num_colors)
{
    const int codeword_length = w * h;

    /* Re-Initialize */
    s->codeword = av_realloc_f(s->codeword, codeword_length, 4 * sizeof(*s->codeword));
    s->codeword_closest_codebook_idxs = av_realloc_f(
        s->codeword_closest_codebook_idxs, codeword_length,
        sizeof(*s->codeword_closest_codebook_idxs));
    s->codebook = av_realloc_f(s->codebook, num_colors, 4 * sizeof(*s->codebook));

    /* build the codeword */
    for (int k = 0, i = 0; i < h; i++)
        for (int j = 0; j < w; j++) {
            s->codeword[k++] = src_data[i * src_linesize + j * 4 + s->b_idx];
            s->codeword[k++] = src_data[i * src_linesize + j * 4 + s->g_idx];
            s->codeword[k++] = src_data[i * src_linesize + j * 4 + s->r_idx];
            s->codeword[k++] = src_data[i * src_linesize + j * 4 + s->a_idx];
        }

    /* compute the codebook */
    avpriv_elbg_do(&s->elbg, s->codeword, 4,
        codeword_length, s->codebook, num_colors, 1,
        s->codeword_closest_codebook_idxs, &s->lfg, 0);

    /* Write Palette */
    for (int i = 0; i < num_colors; i++)
        dst_pal[i] = (s->codebook[i*4+3] << 24) |
                     (s->codebook[i*4+2] << 16) |
                     (s->codebook[i*4+1] <<  8) |
                     (s->codebook[i*4+0] <<  0);

    /* Write Image */
    for (int k = 0, i = 0; i < h; i++)
        for (int j = 0; j < w; j++)
            dst_data[i * dst_linesize + j] =
                s->codeword_closest_codebook_idxs[k++];
}

#define OFFSET(x) offsetof(ASS_Context, x)
#define FLAGS (AV_OPT_FLAG_SUBTITLE_PARAM|AV_OPT_FLAG_FILTERING_PARAM)

static const AVOption text2graphicsub_options[] = {
        { "s",          "output size",                          OFFSET(size),       AV_OPT_TYPE_IMAGE_SIZE,                         .default_val.str = NULL, .flags = FLAGS },
        { "size",       "output size",                          OFFSET(size),       AV_OPT_TYPE_IMAGE_SIZE,                         .default_val.str = NULL, .flags = FLAGS },
        { "n",          "number of output colors",              OFFSET(num_colors), AV_OPT_TYPE_INT,        .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
        { "num_colors", "number of output colors",              OFFSET(num_colors), AV_OPT_TYPE_INT,        .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
        { "f",          "media file from which to load fonts",  OFFSET(filename),   AV_OPT_TYPE_STRING,                             .default_val.str = NULL, .flags = FLAGS },
        { "filename",   "media file from which to load fonts",  OFFSET(filename),   AV_OPT_TYPE_STRING,                             .default_val.str = NULL, .flags = FLAGS },
        { "fd",         "fonts directory",                      OFFSET(fontsdir),   AV_OPT_TYPE_STRING,                             .default_val.str = NULL, .flags = FLAGS },
        { "fontsdir",   "fonts directory",                      OFFSET(fontsdir),   AV_OPT_TYPE_STRING,                             .default_val.str = NULL, .flags = FLAGS },
        { .name =  NULL }
};

static const AVClass text2graphicsub_class = {
        .class_name       = "text2graphicsub",
        .item_name        = av_default_item_name,
        .option           = text2graphicsub_options,
        .version          = LIBAVUTIL_VERSION_INT,
        .category         = AV_CLASS_CATEGORY_FILTER,
};

static const AVFilterPad inputs[] = {
    {
        .name         = "default",
        .type         = AVMEDIA_TYPE_SUBTITLE,
        .filter_frame = filter_frame,
        .config_props = config_input,
    },
};

static const AVFilterPad outputs[] = {
    {
        .name         = "default",
        .type         = AVMEDIA_TYPE_SUBTITLE,
        .config_props = config_output,
    },
};

const AVFilter ff_sf_text2graphicsub = {
        .name            = "text2graphicsub",
        .description     = NULL_IF_CONFIG_SMALL("Convert text subtitles to bitmap subtitles."),
        .init            = init,
        .uninit          = uninit,
        .priv_size       = sizeof(ASS_Context),
        .priv_class      = &text2graphicsub_class,
        FILTER_INPUTS(inputs),
        FILTER_OUTPUTS(outputs),
        FILTER_QUERY_FUNC(query_formats)
};
```

tcoza replied:
Ok, so you mean just replace the prototypes with the definitions. What's wrong with having prototypes?

softworkz replied:
> Ok, so you mean just replace the prototypes with the definitions. What's wrong with having prototypes?

I can't even tell you. No idea. I'm just parroting the things they say and want. In this regard, there was a recent case:

https://www.mail-archive.com/ffmpeg-devel@ffmpeg.org/msg120417.html

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,492 @@
+    ass_set_fonts(context->renderer, NULL, NULL, 1, NULL, 1);
+
+    context->track = NULL;              // Will be init'd in filter_frame
+    context->ass_split_context = NULL;  //
```
Unnecessary. Init is guaranteed to be called just once and the ->priv memory is always zeroed.

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,492 @@
+        AVSubtitleArea *area = frame->subtitle_areas[r];
+        if (area->type == SUBTITLE_BITMAP) continue;
+
+        dialog = avpriv_ass_split_dialog(context->ass_split_context, area->ass);
+        if (track->n_events > 0)
+            ass_free_event(track, 0);
+        track->n_events = 0;
+        ass_alloc_event(track);
+        track->n_events = track->max_events = 1;
+        track->events[0].Start = frame->subtitle_timing.start_pts / (AV_TIME_BASE / 1000);
+        track->events[0].Duration = frame->subtitle_timing.duration / (AV_TIME_BASE / 1000);
+        track->events[0].Effect = av_strdup(dialog->effect);
+        track->events[0].Layer = dialog->layer;
+        track->events[0].MarginL = dialog->margin_l;
+        track->events[0].MarginR = dialog->margin_r;
+        track->events[0].MarginV = dialog->margin_v;
+        track->events[0].Name = av_strdup(dialog->name);
+        track->events[0].Text = av_strdup(dialog->text);
+        track->events[0].ReadOrder = dialog->readorder;
+        track->events[0].Style = 0;
```
This doesn't work out well.  avpriv_ass_split_dialog() is much inferior to the original ass parser and I've seen multiple   errors and problems in testing which are caused by this approach.
Another problem is that you cannot treat each event separate from each other because ass needs to know about all events which are to be shown at a time in order to properly layout/position them. 
With this approach, we get overlapping text blocks.

tcoza replied:
I don't really see how else I can do it, it's sending it to me one event at a time.

softworkz replied:
You need to feed libass in the same way like the other filters do.

Then it will probably also become more clear why I was talking about splitting the result into multiple drawing areas.

tcoza replied:
Which filter? Give me an example of one with libass.

softworkz replied:
Like overlaytextsubs for example:

https://github.com/softworkz/FFmpeg/blob/d4bb52e6b1c5cb82e829ab75fb07d738e9a1313d/libavfilter/vf_overlaytextsubs.c#L551-L556

There are quite a lot of functions in libass for this (like pars_line, etc.). I don't know which would be best for this case and I don't remember why ass_process_chunk() is preferred in the referenced code.

But what's for sure is that **any** libass function for parsing ass lines is better than the ffmpeg-side ass_split_dialog()

tcoza replied:
All right, I'll look into that.

tcoza replied:
Ok, I did that. The tracks render much better now (if a little slower). Even the animations look (almost) great.

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,492 @@
+        avpriv_ass_free_dialog(&dialog);
+
+        image = ass_render_frame(context->renderer, track,
+            track->events[0].Start + track->events[0].Duration / 2, NULL);
```
Picking half-way between start and end can have weird effects on the outputs when animations are active. 
Have you tested files with ass animations?

tcoza replied:
Yes, as you know, I gave up on disabling animations. I've no idea how to do it.

softworkz replied:
Will need to research a bit for the best way of doing this.

tcoza replied:
By the way, I tried not using the midpoint (instead using the start time), it's no good. Some frames get lost. That is because some frames are made to fade in and out, so they are only visible during a interval somewhere around the midpoint. I think the midpoint is pretty good.

softworkz replied:
I know, and I understand why you did it. Hold on I'll respond by e-mail with examples..

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,492 @@
+static const AVOption text2graphicsub_options[] = {
+        { "s",          "output size",                          OFFSET(size),       AV_OPT_TYPE_IMAGE_SIZE,                         .default_val.str = NULL, .flags = FLAGS },
+        { "size",       "output size",                          OFFSET(size),       AV_OPT_TYPE_IMAGE_SIZE,                         .default_val.str = NULL, .flags = FLAGS },
+        { "n",          "number of output colors",              OFFSET(num_colors), AV_OPT_TYPE_INT,        .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
+        { "num_colors", "number of output colors",              OFFSET(num_colors), AV_OPT_TYPE_INT,        .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
```
After testing, I come to the conclusion that 256 is not a useful default.

1. It causes a huge increase in processing time 
2. Like I told already: "Our" quantization is **by far** superior to what the dvdsub encoder does

Just compare the transcoding time and output quality when you set num_colors=4 (when encoding dvd_sub) and you have ass subs with smoothed outlines

tcoza replied:
Yeah, I forgot to change that.

softworkz reviewed (changes requested):
Could you please send me (or provide download link to) two or three files which you have tested with good results?

I tested with 5 files and I haven't seen a single case with acceptable results. There's quite some work to do, I fear. 
If you want me to get involved and help, I can do that, but I'll also respect when you want to finish this on your own and have me provide hints.

It's not about achieving a high level of perfection, but we should still get to a state where someone who tries this with a randomly picked source file will get acceptable results in most cases.

Thanks,
sw

softworkz commented:
PS: You need to do another rebase as I have caught up the current HEAD of ffmpgeg:master. But there shouldn't be any conflicts this time

tcoza commented:
> Could you please send me (or provide download link to) two or three files which you have tested with good results?
> 
> I tested with 5 files and I haven't seen a single case with acceptable results. There's quite some work to do, I fear. If you want me to get involved and help, I can do that, but I'll also respect when you want to finish this on your own and have me provide hints.

Depends what you mean by "acceptable results". To me, what I sent you is acceptable, I don't care that the animations don't look as good as with the original track, I only care about the majority of frames, which are as identical as can be with a 16 color palette.

softworkz commented:
> Depends what you mean by "acceptable results". To me, what I sent you is acceptable, I don't care that the animations don't look as good as with the original track, I only care about the majority of frames, which are as identical as can be with a 16 color palette.

I'll send you some example from the tests I did, I think this will be the best so you can see the individual problems I'm talking about.

tcoza commented:
All right

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,441 @@
+    {
+        ASS_Image *image;
+        AVSubtitleArea *area = frame->subtitle_areas[r];
+        if (area->type == SUBTITLE_BITMAP) continue;
```
```
if (area->type != AV_SUBTITLE_FMT_ASS) continue;
```

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,441 @@
+        }
+
+        ass_image_to_area_palletization(context, image, area);
+        area->type = SUBTITLE_BITMAP;
```
AV_SUBTITLE_FMT_BITMAP

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,441 @@
+    const int64_t duration   = av_rescale_q(frame->subtitle_timing.duration, AV_TIME_BASE_Q, av_make_q(1, 1000));
+    
+    if (!context->track) {      // Init track
+        if (frame->subtitle_header)
+            context->track = ass_read_memory(context->library,
+                (char *) frame->subtitle_header->data,
+                frame->subtitle_header->size, NULL);
+        else context->track = ass_new_track(context->library);
```
Please copy the logic with process_header() from overlaytextsubs here. This is probably the reason why some things look different than with VLC player (and overlaytextsubs and subtitles filter)

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,441 @@
+                frame->subtitle_header->size, NULL);
+        else context->track = ass_new_track(context->library);
+    }
+
```
After the header reading, you need to add this code right here:

```
    if (frame->repeat_sub)
        return 0;
```

softworkz reviewed (changes requested):
Added some more notes on today's code update

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,441 @@
+        }
+
+        ass_image_to_area_palletization(context, image, area);
+        area->type = SUBTITLE_BITMAP;
```
and this must go to the top of the loop. We need to make sure that all areas are bitmap type

softworkz replied:
oh - no, we must not emit any empty areas either.

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,441 @@
+        else context->track = ass_new_track(context->library);
+    }
+
+    for (int r = 0; r < frame->num_subtitle_areas; r++)
```
Before this, you need to add:

```
    ret = av_frame_make_writable(frame);
    if (ret < 0) return ret;

```

softworkz committed:
fftools/ffmpeg: Introduce subtitle filtering and new frame-based subtitle encoding

This commit actually enables subtitle filtering in ffmpeg by
sending and receiving subtitle frames to and from a filtergraph.

The heartbeat functionality from the previous sub2video implementation
is removed and now provided by the 'subfeed' filter.
The other part of sub2video functionality is retained by
auto-insertion of the new graphicsub2video filter.

Justification for changed test refs:

- sub2video
  The previous results were incorrect. The command line for this test
  specifies -r 5 (5 frames per second), which is now fulfilled by the
  additional frames in the reference output.
  Example: The first subtitle time is 499000, the second is 15355000,
  which means 0.5s and 15.35s with a difference of 14.85s.
  15s * 5fps = 75 frames and that's now the exact number of video
  frames between these two subtitle events.

- sub2video_basic
  The previous results had some incorrect output because multiple
  frames had the same dts
  The non-empty content frames are visually identical, the different
  CRC is due to the different blending algorithm that is being used.

- sub2video_time_limited
  Subtitle frames are emitted to the filter graphs at a 5 fps rate
  by default. The time limit for this test is 15s * 5fps = 75 frames
  which matches the count in the new ref.

- sub-dvb
  Running ffprobe -show_frames on the source file shows that there
  are 7 subtitle frames with 0 rects in the source at the start
  and 2 at the end. This translates to the 14 and 4 additional
  entries in the new test results.

- filter-overlay-dvdsub-2397
  Overlay results have slightly different CRCs due to different
  blending implementation

- sub-scc
  The first entry is no longer in the output because it is before
  the actual start time and the strim filter removes such entries
  now (like for video and audio)

Signed-off-by: softworkz <softworkz@hotmail.com>

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,526 @@
+        return AVERROR_EXIT;
+    }
+    ass_set_frame_size(context->renderer, context->size.width, context->size.height);
+    ass_set_storage_size(context->renderer, context->size.width, context->size.height);
```
We need to change this line to

```
    ass_set_storage_size(context->renderer, inlink->w, inlink->h);
```

This must be the input size, which is the size for which the ASS subs were originally created for.
libass uses this to properly rescale values like font size, outline size, shadow depths and other things to preserve the appearance even the target size is different from the original size.

softworkz commented on libavfilter/sf_text2graphicsub.c:
```diff
@@ -0,0 +1,526 @@
+    av_log(NULL, AV_LOG_DEBUG, "successfully rendered ass: %s\n", area->ass);
+
+    frame->format = AV_SUBTITLE_FMT_BITMAP;
+    return ff_filter_frame(outlink, frame);
```
Here we need to add this:

```
    frame->width = context->size.width;
    frame->height = context->size.height;
```

softworkz commented:
Hi, I just pushed another base update (probably the last one). When you rebase and make the two changes, then I think we can make the first submission with your filter included. 
Please let me know when you're ready or whether there are any concerns left on our side.
Thanks,
sw

tcoza commented:
Hey, now that you've approved the pull request, do I still force-push to the same commit, or do I create other commits?

tcoza commented:
Yo, is this roughly how the new stripstyles function is used? 
![image](https://user-images.githubusercontent.com/4213839/174940390-0266986f-a68b-4de6-adc7-6dc040cb1261.png)

In that case, there might be a bug in your implementation. For most lines it works ok, but take a look at this output:

```
Before: {\fad(1500,1200)\pos(960,360)\blur6}{\fs78}Code Geass\N{\fs39}- Lelouch of the Rebellion -
After:  {\pos(960,360)}{\fs78}Code Geass\NCode Geass{\fs39}Code Geass\N{\fs39}- Lelouch of the Re
```

softworkz commented:
> Hey, now that you've approved the pull request, do I still force-push to the same commit, or do I create other commits?

You still force-push that single commit.

Once I have submitted the first version, your commit will be included in my branch (with your signature and ownership).

For subsequent updates, you will create a new PR with a single commit. When we publish another new version, I will then merge your PR, locally squash your new commit with the previous commit (so your filter will still be a single commit) before submitting the new version (and so on..)

softworkz commented:
> Yo, is this roughly how the new stripstyles function is used? ![image](https://user-images.githubusercontent.com/4213839/174940390-0266986f-a68b-4de6-adc7-6dc040cb1261.png)
> 
> In that case, there might be a bug in your implementation. For most lines it works ok, but take a look at this output:
> 
> ```
> Before: {\fad(1500,1200)\pos(960,360)\blur6}{\fs78}Code Geass\N{\fs39}- Lelouch of the Rebellion -
> After:  {\pos(960,360)}{\fs78}Code Geass\NCode Geass{\fs39}Code Geass\N{\fs39}- Lelouch of the Re
> ```

What are you doing in the callbacks? I'm handling the text callback to check whether there is text present at all in the line, and if not, the whole line will be set to null.

softworkz commented:
I will check your example, though´.

tcoza commented:
> What are you doing in the callbacks? I'm handling the text callback to check whether there is text present at all in the line, and if not, the whole line will be set to null.

Well as you see, they're all null. I thought we had no need for callbacks anymore. But this isn't a bad idea, tho.

softworkz commented:
> > What are you doing in the callbacks? I'm handling the text callback to check whether there is text present at all in the line, and if not, the whole line will be set to null.
> 
> Well as you see, they're all null. I thought we had no need for callbacks anymore. But this isn't a bad idea, tho.

This could be discussed. I just didn't want to do too much modifications of the ass_split implementation because it is used by many subtitle encoders.

tcoza commented:
Looks better. The implementation is still buggy tho, since I'm getting warnings such as this:
```
failed to render ass: 627,0,Romaji,,0,0,0,Effector [fx],{\an5\move(1341.3,50.4,1521.3,50.4)\1c&HFFFFFF&\3c&H5747F2&\4c&H454545&\fscx55\fscy55\bord2.2\blur4.5\shad0\fad(0,100)\fr-3124.8\t(\fr-3484.8\fscx0\fscy0)\t(0,80,\bord4.5\3c&HFFFFFF&)\t(80,160,\bord2.2\3c&H5747F2&)\t(160,240,\bord4.5\3c&HFFFFFF&)\t(240,320,\bord2.2\3c&H5747F2&)\t(320,400,\bord4.5\3c&HFFFFFF&)\p1}m 75 0 b 33 0 0 33 0 75 b 0 117 33 150 75 150 b 117 150 150 117 150 75 b 150 33 117 0 75 0 m 145.5 75 b 145.5 114 114 145.5 75 145.5 b 36 145.5 4.5 114 4.5 75 b 4.5 37.5 36 4.5 75 4.5 l 75 12 b 97.5 12 120 25.5 130.5 46.5 l 136.5 42 b 141 48 145.5 57 145.5 75 m 13.5 130.5 l 16.5 133.5 l 30 120 l 27 117 l 13.5 130.5
```
containing tags which were supposed to be stripped. But that line I sent you before is fixed at least.

softworkz commented:
> Looks better. The implementation is still buggy tho, since I'm getting warnings such as this:
> 
> ```
> failed to render ass: 627,0,Romaji,,0,0,0,Effector [fx],{\an5\move(1341.3,50.4,1521.3,50.4)\1c&HFFFFFF&\3c&H5747F2&\4c&H454545&\fscx55\fscy55\bord2.2\blur4.5\shad0\fad(0,100)\fr-3124.8\t(\fr-3484.8\fscx0\fscy0)\t(0,80,\bord4.5\3c&HFFFFFF&)\t(80,160,\bord2.2\3c&H5747F2&)\t(160,240,\bord4.5\3c&HFFFFFF&)\t(240,320,\bord2.2\3c&H5747F2&)\t(320,400,\bord4.5\3c&HFFFFFF&)\p1}m 75 0 b 33 0 0 33 0 75 b 0 117 33 150 75 150 b 117 150 150 117 150 75 b 150 33 117 0 75 0 m 145.5 75 b 145.5 114 114 145.5 75 145.5 b 36 145.5 4.5 114 4.5 75 b 4.5 37.5 36 4.5 75 4.5 l 75 12 b 97.5 12 120 25.5 130.5 46.5 l 136.5 42 b 141 48 145.5 57 145.5 75 m 13.5 130.5 l 16.5 133.5 l 30 120 l 27 117 l 13.5 130.5
> ```
> 
> containing tags which were supposed to be stripped. But that line I sent you before is fixed at least.

Did you apply the patch that I sent you?

tcoza commented:
Wait, I applied the patch and it doesn't give those messages anymore. But now it seems to be stripping too many styles. Some frames seem to be completely removed. It's just like the stripstyles filter.

Should I send you an example?

softworkz commented:
> Wait, I applied the patch and it doesn't give those messages anymore. But now it seems to be stripping too many styles. Some frames seem to be completely removed. It's just like the stripstyles filter.

Can you give an example what gets removed but shouldn't?

softworkz commented:
> Should I send you an example?

LOL 
(a concurrency issue)

tcoza commented:
nice

tcoza commented:
Ok I am sending an episode I use for testing. The opening is supposed to have subtitles are the top, and it did even with the stripstyles option, but it doesn't anymore since the latest patch. You only see flashes of vertical letters sometimes. This is exactly what happened when I used the stripstyles filter.

softworkz commented:
OK, I'll look at it.

BTW, did you watch the demo video I sent to you (link)?

tcoza commented:
Ok it's on the repo.
Not sure. I saw some screenshots. Are you talking about SubsTest_2_H264_8bit.mkv?

softworkz commented:
No, the 1drv.ms link I sent via e-mail

softworkz commented:
Watch!

tcoza commented:
Yeah, looks dope.

softworkz commented:
Done with a single ffmpeg command.
This demo is what I wanted to do for a long time. I just hadn't time to do a t2gs filter..

tcoza commented:
Wow, that's pretty good. Glad I could be of help. By the way, big thanks for helping me through this, you've been very kind.

softworkz commented:
I hate like many of the ffmpeg devs behave. It's all based on negativism. The best you can get is silence, which means nobody objects, and below that level, there's a huge bandwidth from justified objections to biased objections - rarely objections are coming with alternate suggestions - to "talking-to-death" behavior and finally even down to personal insults. 

It's nothing for the faint-of-heart, but I'm standing strong as well - for and behind the "Subtitle Filtering" addition, which is surely a major one (and it's not coming from one of their core developers, which is an irritating factor for them). This has already gone through about 30 iterations by now, and most objections have been ruled out already.
For the finale now, I am very glad to have you on board which increases the impact,  importance and standing of this submission, so thanks for joining and trusting me!

softworkz commented:
So, for that final issue, I don't have immediate good news.

But first of all - let's put all concerns in terms of making sense or being reasonable aside and see what we get in all possible cases.

## [E01_a.MKV] With stripstyles=0

This is probably what you mean to have seen before:

![image](https://user-images.githubusercontent.com/4985349/175455462-98336b75-b724-400d-bdeb-c2464f7ea854.png)

## [E01_b.MKV] With stripstyles=1 but not dropping text subject to animation

(I have temporarily added an option of this)

![image](https://user-images.githubusercontent.com/4985349/175455685-7a8d5211-cd92-4937-b3da-770698ca3312.png)

## [E01_c.MKV]  Normal Operation with stripstyles=1

**=> No lyrics visible**

softworkz commented:
## Now let's compare the output files from all three cases

### Format

![image](https://user-images.githubusercontent.com/4985349/175462538-995b9db1-8282-47a7-85f6-becb626bca3b.png)

### Video Streams

All identical information

![image](https://user-images.githubusercontent.com/4985349/175462658-0a0d63f4-fd92-480d-ad5a-0c5ae5880e40.png)


### Subtitle Streams

All identical information as well

![image](https://user-images.githubusercontent.com/4985349/175462771-8f5cb395-8b74-4e2c-aa68-203b174a949b.png)

but....

### Frames Counts

![image](https://user-images.githubusercontent.com/4985349/175463292-8cc13e43-e564-43b5-ba4c-490c2bbd87df.png)


![image](https://user-images.githubusercontent.com/4985349/175463379-c893e697-0d27-43ac-a4f2-0a703ad98a53.png)


![image](https://user-images.githubusercontent.com/4985349/175463587-dc1b4ab7-f020-4d21-915a-746029267612.png)

softworkz commented:
The B-file is larger than the A-file because the bitmaps created are larger.

### Bottom Line 1

Neither A nor B is an acceptable result file.

2k subtitle bitmaps for 60k video  frames is insane.  VLC may play it, but normal DVD or DVB devices won't do it.

### Bottom Line 2

If you still want to produce this kind of output, you are better off by totally disabling stripstyles. A combination of stripstyles but keeping the text doesn't make it any better - actually even worse: larger bitmaps + worse appearance

### Possible Solution

would be to de-couple input and output of the filter and have a fixed-framerate output query on the filter (like textsub2video does). But for showing animations, you would still need a certain frame rate (maybe 5 fps) and when an animation is ongoing, you will have a different bitmap(-set) every 200ms. And bitmap subs are just not made for this.

So, IMO, it's not worth the effort. And when you want to have it for fun working in VLC, you can still disable stripstyles.

softworkz commented:
BTW, it will not be like this for all videos. Some have a regular base text for the lyrics and the animations are decorating these on top.

In your specific video, every single letter is a separate ASS event with animation tags - hence there is no easy way to get a useful output.

![image](https://user-images.githubusercontent.com/4985349/175464996-33cab98f-d742-4365-a900-be5514914fad.png)

softworkz commented:
I noticed a tiny parsing bug which I have fixed and pushed, so you would need to rebase once again and push the complete commit.
I'd be ready to submit it - are you?

tcoza commented:
I don't care that much about the opening subtitles anyway, so it's fine. I just thought it was weird how it used to work, but not anymore.

Sure, let's submit.

softworkz commented:
Done!

I'm closing as I've cherry-picked your commit directly from your branch.
